### ISSUE TITLE
Optimization: Compile and link shaders asynchronously.

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4287]
This task is automatically imported from the old Task Issue Board and it was originally created by Keziah Wesley.
Original issue is [here](https://www.pivotaltracker.com/story/show/181681597).
<hr />

This task was dropped and continued here: https://www.pivotaltracker.com/story/show/181296538

We are currently blocking to wait for GL shaders to compile. This is an action the browser performs in the background; we could substantially improve performance by using a non-blocking interface to poll for compilation asynchronously.

The blocking occurs in [system::gpu::shader::compile_shader](https://github.com/enso-org/enso/blob/b9172b46a170e0a1bacd52be403e12eca2af5287/lib/rust/ensogl/core/src/system/gpu/shader.rs#L131), specifically [checking compilation status](https://github.com/enso-org/enso/blob/b9172b46a170e0a1bacd52be403e12eca2af5287/lib/rust/ensogl/core/src/system/gpu/shader.rs#L104). It looks like everything is already structured with the intent that shaders can be compiled asynchronously, it's just being defeated by eager status checks. If we instead use the `KHR_parallel_shader_compile` API to poll compilation status, we should never need to block the main thread to wait for shaders to compile.

## Comments:
	Plan: Apply 3 best practices from here: https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices
- "Compile Shaders and Link Programs in parallel"
- "Prefer KHR_parallel_shader_compile"
- "Don't check shader compile status unless linking fails"

Implementation:
- `compile_program` (which is run during Scene::update) should add the program to a compile-queue and accept a callback.
- At the end of each Scene::update, we go over the queue and start compiling all the shaders; then we go over all the shader pairs, and request they be linked into programs.
- At the beginning of each Scene::update, poll the status of each outstanding link operation, and if it is done run its callback or report its error. (We check readiness at the beginning of Scene::update because later in that function is where programs are used).
- (We must ensure if compilations complete out of order, then when the already-outdated program becomes ready it is ignored.)
- (We should retry a compilation if it fails due to lost context, but limit the number of attempts in case we are losing the context because compiling the shader is crashing the GPU process or something.) (Keziah Wesley - Mar 25, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-03-29):

**Progress:** Short day. Sketched out the structure of the new async compile pipeline. It should be finished by 2022-04-05.

**Next Day:** Next day I will be working on the [##181681597 task](https://www.pivotaltracker.com/story/show/181681597). Start refactoring the current shader compilation code into the new structure. (Enso Bot - Mar 30, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-03-30):

**Progress:** Implemented the happy path; it's working, and the performance benefits are actually greater than expected. Will probably finish this earlier than originally estimated because some follow-on work turns out not to be as important as I thought it would be, so I'll split it into a separate task we might address later. It should be finished by 2022-04-03.

**Next Day:** Next day I will be working on the [##181681597 task](https://www.pivotaltracker.com/story/show/181681597). Implement error handling for the new compilation pipeline. (Enso Bot - Mar 30, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-03-31):

**Progress:** Roughly finished the async shader compile pipeline. Needs testing and a little simplification. It should be finished by 2022-04-03.

**Next Day:** Next day I will be working on the [##181610155 task](https://www.pivotaltracker.com/story/show/181610155). Look in to stack overflow in a test in define_id branch (Enso Bot - Apr 2, 2022)
<hr />
Closing this as it was subsumed into Wojciech's ongoing work. (Keziah Wesley - Apr 5, 2022)
<hr />
Reopenning to track my progress. (Wojciech Daniło - Apr 6, 2022)
<hr />
**Keziah Wesley** reports a new **🔴 DELAY** for today (2022-04-06):

**Summary:** There is **3 days** delay in implementation of the [Optimization: Compile and link shaders asynchronously. (#181681597) task](https://www.pivotaltracker.com/story/show/181681597).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Not a delay. Did some review/bug investigation for task I finished working on last week. (Enso Bot - Apr 7, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-04-06):

**Progress:** Reviewed async compilation PR; found a performance bug due to undocumented Chromium behavior; worked out a solution. It should be finished by 2022-04-06.

**Next Day:** Next day I will be working on the [##181775808 task](https://www.pivotaltracker.com/story/show/181775808). Get back to batch mode: finish implementation by defining some basic debug-workflows. (Enso Bot - Apr 7, 2022)
<hr />
